### DISCUSSION TITLE
Allows to include test files in the `pages` directory

### DISCUSSION THREAD

**Discussion by vicasas** [vercel/next.js#51109] [Ideas]
### Goals

1. Allow adding an exclude pattern for the `pageExtensions` option of the `next.config.js` file.
2. Or make the `.{test,spec}.{js,jsx,ts,tsx}` extensions supported by default inside the `page` directory.

### Background

This function is required in order to support the test files inside the `pages` directory.

If there is a test file inside the directory when executing `npm run dev` it continues to work normally, but when executing `npm run build` the compilation breaks due to the presence of the file.

There is currently a `pageextensions` option in the `next.config.js` file where you can control this problem. But for this it is necessary to modify the names of the files that are pages, adding a `.page` prefix so as not to clash with `.test` or `.spec` files.

I think you shouldn't have to rename the pages in order to avoid the error with the test files. Perhaps this could be improved by adding an exclude pattern.

They could also include these extensions by default, since they are the most common for tests.

### Proposal

1.

To improve the development experience, let's add an exclude pattern to do this:

```js
module.exports = {
  pageExtensions: ['!.test.tsx'],
}
```

Instead of this:

```js
module.exports = {
  pageExtensions: ['page.tsx', 'page.ts', 'page.jsx', 'page.js'],
}
```

2.

Default that allows files with `.test.{js,jsx,ts,tsx}` or `.spec.{js,jsx,ts,tsx}` extensions to be added without causing a compilation error.

**Comment by harrisondeo** [comment]
I completely agree with this, Im currently working in a repo where we are using Module Federation and renaming the pages files to '.pages.tsx' breaks the Module federation package. It would be nice to be able to "exclude these file extensions" rather than force page name changes
